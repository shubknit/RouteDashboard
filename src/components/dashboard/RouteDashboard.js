import React, { Component } from 'react';
import { Header } from '../header/Header';
import { UserForm } from '../form/UserForm';
import  { ViewRoute }  from '../direction/ViewRoute';
import { getRouteDetails } from '../../services/routeapi/RouteDetailsAPI';
import { responseError } from '../../constants/apiConfig';

// This Component acts a dashboard  which includes user form and map view.
export class RouteDashboard extends Component {
	constructor() {
		super();
		this.state = {
			mapData : {},
			isFetching: false
		}
  }
	// setstate when call turned into server error from API
	handleError = () => {
		this.setState({
			status: responseError.status,
			error: responseError.error,
			isFetching: false
		})
	}

	// set state when call get success from API
	handleResponse = (response) => {
		if(response && response.error){
			this.setState({
	  		status: response.status,
				error: response.error,
				isFetching: false
			})
		}
		else if(response && response.path){
			this.setState({
				mapData: {
					start: response.path[0],
					waypoint: response.path[1],
					end: response.path[2],
					distance: response.total_distance,
					time: response.total_time
				},
				status: response.status,
				isFetching: false
			})
		}
	}
  
  // callback when form submit is called with start and end point.	
  addLocation = async (data) => {
		const {start, end} = data;
		this.setState({
			mapData: {},
			error: '',
			status: '',
			isFetching: true,
		})
		const response = await getRouteDetails(start, end).catch(e => {
			this.handleError();
		});
		this.handleResponse(response);	
	}

	// Reset the map, form and button state when reset button is clicked
	resetPage = () => {
		this.setState({
			mapData : {},
			isFetching: false,
			error: '',
			status: ''
		})
	}
	
	render(){
		return (
			<div className = 'main-container'>
				<Header/>
				<div className = 'content'>
            <UserForm  addLocation = {this.addLocation} data = {this.state} resetForm = {this.resetPage} />
					<div className = 'map-container'>
						<ViewRoute data = {this.state}/>
					</div>
				</div>
			</div>
		)
	}

}