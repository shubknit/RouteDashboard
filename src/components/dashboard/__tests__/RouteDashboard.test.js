import React from 'react';
import { RouteDashboard } from '../RouteDashboard';
import renderer from 'react-test-renderer';



it('should render the dashboard page correctly', () => {
    const wrapper = renderer.create(<RouteDashboard/>).toJSON();
    expect(wrapper).toMatchSnapshot();
})