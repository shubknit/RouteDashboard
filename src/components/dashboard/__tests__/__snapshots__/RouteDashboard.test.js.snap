// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the dashboard page correctly 1`] = `
<div
  className="main-container"
>
  <header
    className="header"
  >
    <div
      className="content-container"
    >
      <div
        className="header__content"
      >
        <h1>
          Router Finder App
        </h1>
      </div>
    </div>
  </header>
  <div
    className="content"
  >
    <div
      className="form-container"
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <label
          htmlFor="start-point"
        >
           Starting Location 
        </label>
        <input
          className="text-input"
          id="start-point"
          placeholder="start point"
          required={true}
          type="text"
        />
        <label
          htmlFor="end-point"
        >
           Drop-off point
        </label>
        <input
          className="text-input"
          id="end-point"
          placeholder="end point"
          required={true}
          type="text"
        />
        <div
          className="route-info"
        />
        <div
          className="button-container"
        >
          <button
            className="button"
            type="submit"
          >
             
            Submit
             
          </button>
          <button
            className="button"
            onClick={[Function]}
            type="reset"
          >
             Reset 
          </button>
        </div>
      </form>
    </div>
    <div
      className="map-container"
    >
      <div
        id="map"
      />
    </div>
  </div>
</div>
`;
